### PR TITLE
Update test_duplicate_feature

### PR DESCRIPTION
**Pull request by jgersak**
split into multiple tests
add parametrization
add test for duplicate deep
add types

### CODE REVIEW

File: core/eolearn/tests/test_core_tasks.py [hunk 1]
```diff
@@ -172,47 +173,45 @@ def test_add_rename_remove_feature(patch):
...
+        [(FeatureType.DATA, "bands", f"bands{i}") for i in range(5)],
+    ],
+)
+def test_duplicate_feature(feature_specification: List[FeaturesSpecification], patch: EOPatch) -> None:
+    patch = DuplicateFeatureTask(feature_specification)(patch)
 
-    patch = AddFeatureTask((FeatureType.MASK, feature_name))(patch, mask_data)
+    for f_type, f_name, f_dup_name in feature_specification:
+        assert f_dup_name in patch[f_type]
+        assert id(patch[(f_type, f_name)]) == id(patch[(f_type, f_dup_name)])
```
Comment (zigaLuksic) [inline core/eolearn/tests/test_core_tasks.py:189]: add assertion message why this must hold

File: core/eolearn/tests/test_core_tasks.py [hunk 2]
```diff
@@ -172,47 +173,45 @@ def test_add_rename_remove_feature(patch):
...
+    duplicate_task_deep = DuplicateFeatureTask((FeatureType.DATA, "bands", "bands_dup_deep"), deep_copy=True)
     patch = duplicate_task(patch)
+    patch = duplicate_task_deep(patch)
+
+    assert all(f_name in patch.data for f_name in ["bands_dup", "bands_dup_deep"])
+    assert_array_equal(patch.data["bands_dup"], patch.data["bands_dup_deep"])
+    assert id(patch.data["bands_dup"]) != id(patch.data["bands_dup_deep"])
+    patch.data["bands"] += 1
+    assert_array_equal(patch.data["bands_dup"], patch.data["bands"])
+    assert not np.array_equal(patch.data["bands_dup"], patch.data["bands_dup_deep"])
```
Comment (zigaLuksic) [inline core/eolearn/tests/test_core_tasks.py:204]: this looks like it tests too much
why are you testing shallow duplication as well? isn't that covered in the previous test?

File: core/eolearn/tests/test_core_tasks.py [hunk 3]
```diff
@@ -172,47 +173,45 @@ def test_add_rename_remove_feature(patch):
...
-    assert np.array_equal(
-        patch.mask[duplicate_name], mask_data
-    ), "Feature was not duplicated correctly. Data does not match."
+def test_duplicate_feature_fails(patch: EOPatch) -> None:
+    duplicate_task = DuplicateFeatureTask((FeatureType.DATA, "bands", "bands_dup"))
+    patch = duplicate_task(patch)
 
     with pytest.raises(ValueError):
         # Expected a ValueError when creating an already exising feature.
         patch = duplicate_task(patch)
```
Comment (zigaLuksic) [inline core/eolearn/tests/test_core_tasks.py:213]: no reason to call it twice, you know what the names in the patch fixture are, use that as input

Comment (jgersak) [reply core/eolearn/tests/test_core_tasks.py]: done

File: core/eolearn/tests/test_core_tasks.py (hunk 2 above)
Comment (jgersak) [reply core/eolearn/tests/test_core_tasks.py]: agree updated

File: core/eolearn/tests/test_core_tasks.py (hunk 1 above)
Comment (jgersak) [reply core/eolearn/tests/test_core_tasks.py]: done

File: core/eolearn/tests/test_core_tasks.py [hunk 4]
```diff
@@ -199,46 +199,38 @@ def test_remove_fails(patch: EOPatch) -> None:
...
+def test_duplicate_feature(feature_specification: List[FeaturesSpecification], patch: EOPatch) -> None:
+    patch = DuplicateFeatureTask(feature_specification)(patch)
 
-    with pytest.raises(ValueError):
-        # Expected a ValueError when creating an already exising feature.
-        patch = duplicate_task(patch)
+    for f_type, f_name, f_dup_name in feature_specification:
+        assert f_dup_name in patch[f_type]
+        assert id(patch[(f_type, f_name)]) == id(patch[(f_type, f_dup_name)]), "DuplicateFeatureTask used deep copy."
+        assert_array_equal(patch[(f_type, f_name)], patch[(f_type, f_dup_name)])
```
Comment (zigaLuksic) [inline core/eolearn/tests/test_core_tasks.py:216]: this shouldn't be equal, as you checked that it is the same object

File: core/eolearn/tests/test_core_tasks.py [hunk 5]
```diff
@@ -199,46 +199,38 @@ def test_remove_fails(patch: EOPatch) -> None:
...
+)
+def test_duplicate_feature(feature_specification: List[FeaturesSpecification], patch: EOPatch) -> None:
+    patch = DuplicateFeatureTask(feature_specification)(patch)
 
-    with pytest.raises(ValueError):
-        # Expected a ValueError when creating an already exising feature.
-        patch = duplicate_task(patch)
+    for f_type, f_name, f_dup_name in feature_specification:
+        assert f_dup_name in patch[f_type]
+        assert id(patch[(f_type, f_name)]) == id(patch[(f_type, f_dup_name)]), "DuplicateFeatureTask used deep copy."
```
Comment (zigaLuksic) [inline core/eolearn/tests/test_core_tasks.py:215]:
```suggestion
        assert id(patch[(f_type, f_name)]) == id(patch[(f_type, f_dup_name)]), "Data was deep-copied."
```

File: core/eolearn/tests/test_core_tasks.py [hunk 6]
```diff
@@ -199,46 +199,38 @@ def test_remove_fails(patch: EOPatch) -> None:
...
+    patch = DuplicateFeatureTask((FeatureType.DATA, "bands", "bands_dup"), deep_copy=True)(patch)
 
-    patch = DuplicateFeatureTask((FeatureType.MASK, "MASK1", "DEEP"), deep_copy=True)(patch)
-    assert id(patch.mask["MASK1"]) != id(patch.mask["DEEP"])
-    assert np.array_equal(
-        patch.mask["MASK1"], patch.mask["DEEP"]
-    ), "Feature was not duplicated correctly. Data does not match."
+    assert "bands_dup" in patch.data
+    assert_array_equal(patch.data["bands_dup"], patch.data["bands"])
+    assert id(patch.data["bands_dup"]) != id(patch.data["bands"])
```
Comment (zigaLuksic) [inline core/eolearn/tests/test_core_tasks.py:224]: either this line is not needed, or the following two (where you change one and check that the other is changed) are not needed

File: core/eolearn/tests/test_core_tasks.py (hunk 4 above)
Comment (mlubej) [reply core/eolearn/tests/test_core_tasks.py:216]: Do you mean _should_, and that this check is therefore redundant?

Comment (jgersak) [reply core/eolearn/tests/test_core_tasks.py:217]: Updated

Comment (zigaLuksic) [reply core/eolearn/tests/test_core_tasks.py:217]: ah yes, words are hard

File: core/eolearn/tests/test_core_tasks.py [hunk 7]
```diff
@@ -199,46 +199,28 @@ def test_remove_fails(patch: EOPatch) -> None:
...
+)
+@pytest.mark.parametrize("deep", [True, False])
+def test_duplicate_feature(feature_specification: List[FeaturesSpecification], deep: bool, patch: EOPatch) -> None:
+    patch = DuplicateFeatureTask(feature_specification, deep)(patch)
 
-    duplicate_task = DuplicateFeatureTask((FeatureType.MASK, feature_name, duplicate_name))
-    patch = duplicate_task(patch)
+    for f_type, f_name, f_dup_name in feature_specification:
+        assert f_dup_name in patch[f_type]
+        assert (id(patch[(f_type, f_name)]) == id(patch[(f_type, f_dup_name)])) != deep
```
Comment (zigaLuksic) [inline core/eolearn/tests/test_core_tasks.py:216]: no

if you want to have both in the same test, then do something like
```python
original_id = id(patch[(ftype, fname)])
duplicated_id = id(patch[(ftype, fname)])
assert original_id != duplicated_id if deep else original_id == duplicated_id
```

Comment (jgersak) [reply core/eolearn/tests/test_core_tasks.py]: done

File: core/eolearn/tests/test_core_tasks.py [hunk 8]
```diff
@@ -199,46 +199,30 @@ def test_remove_fails(patch: EOPatch) -> None:
...
+    patch = DuplicateFeatureTask(feature_specification, deep)(patch)
 
-    duplicate_task = DuplicateFeatureTask((FeatureType.MASK, feature_name, duplicate_name))
-    patch = duplicate_task(patch)
+    for f_type, f_name, f_dup_name in feature_specification:
+        assert f_dup_name in patch[f_type]
+        original_id = id(patch[(f_type, f_name)])
+        duplicated_id = id(patch[(f_type, f_dup_name)])
+        assert original_id != duplicated_id if deep else original_id == duplicated_id
+        assert_array_equal(patch[(f_type, f_name)], patch[(f_type, f_dup_name)])
```
Comment (zigaLuksic) [inline core/eolearn/tests/test_core_tasks.py:219]: just another 'prettification' suggestion.
I would do `orig_feat = f_type, f_name` and `dup_feat = f_type, f_dup_name` (or use longer names `original_feature` if it makes the code prettier).
You can then do `assert duplicated_feature in patch` and you get rid of a lot of parenthesis later on (and it is more clear whether you're looking for the original or duplicate)

File: core/eolearn/tests/test_core_tasks.py (hunk 7 above)
Comment (jgersak) [reply core/eolearn/tests/test_core_tasks.py]: done

File: core/eolearn/tests/test_core_tasks.py (hunk 8 above)
Comment (jgersak) [reply core/eolearn/tests/test_core_tasks.py]: done

**Review by zigaLuksic — APPROVED** [review]
